src/Npgsql/ConnectorPool.cs

```diff
@@ -320,20 +328,22 @@ await Task.WhenAny(tcs.Task, Task.Delay(timeLeft, delayCancellationToken.Token))
                                 }
                                 else
                                 {
-                                    using (cancellationToken.Register(() => tcs.SetCanceled()))
+                                    using (cancellationToken.Register(s => ((TaskCompletionSource<NpgsqlConnector?>)s).SetCanceled(), tcs))
                                         await tcs.Task;
                                 }
                             }
                             else
                             {
                                 if (timeout.IsSet)
                                 {
-                                    var timeLeft = timeout.TimeLeft;
-                                    if (timeLeft <= TimeSpan.Zero || !tcs.Task.Wait(timeLeft))
+                                    var timeLeft = timeout.TimeLeft.TotalMilliseconds;
+                                    if (timeLeft > int.MaxValue)
```

Comment:
Thanks... So if my calculations are correct we only support around 24 days of timeout - let's see if someone complains :)